### PR TITLE
feat: expose url field on issue api.

### LINKED ISSUE #402: [API] Issue and pull-request should expose `url`
## Description

Since Gitea aims to be 100% GitHub-compatible, issues should expose an `url` field in the API. This field is the API URL that can be used to retrieve the given issue/PR.

Reference:

- GitHub issue API: https://developer.github.com/v3/issues/#get-a-single-issue
- GitHub pull-request API: https://developer.github.com/v3/pulls/#get-a-single-pull-request

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/40132214-api-issue-and-pull-request-should-expose-url?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by appleboy**
fix https://github.com/go-gitea/gitea/issues/402

### CODE REVIEW

**Comment by appleboy** [comment]
This would be better include in 1.1.0 release.

**Comment by lunny** [comment]
Let's decided it before release

**Comment by appleboy** [comment]
@lunny OK

**Comment by lunny** [comment]
trusted LGTM

**Comment by lunny** [comment]
@rymai @bkcsoft could you confirm this?

File: models/issue.go [hunk 1]
```diff
@@ -246,6 +246,7 @@ func (issue *Issue) APIFormat() *api.Issue {
 
 	apiIssue := &api.Issue{
 		ID:       issue.ID,
+		URL:      issue.HTMLURL(),
```
Comment (rymai) [inline models/issue.go]: I'm not sure about that. From GitHub's API it seems that it should be the API URL to get the issue, not the HTML URL: https://developer.github.com/v3/issues/#get-a-single-issue

```
"url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
```

Comment (appleboy) [reply models/issue.go]: Didn't we define the API URL for Gitea? cc @lunny

Comment (bkcsoft) [inline models/issue.go]: Indeed it is the API-URL :) and that function doesn't exist, so we need a `issue.APIURL()` as well (should be fairly simple)

Comment (bkcsoft) [inline models/issue.go]: Indeed it is the API-URL :) and that function doesn't exist, so we need a `issue.APIURL()` as well (should be fairly simple)

Comment (bkcsoft) [reply models/issue.go]: Indeed it is the API-URL :) and that function doesn't exist, so we need a `issue.APIURL()` as well (should be fairly simple)

Comment (lunny) [reply models/issue.go]: @appleboy It seems there is not.

Comment (appleboy) [reply models/issue.go]: But we didn't have API URL in `app.ini`, we only have `ROOT_URL` config now.

Comment (appleboy) [reply models/issue.go]: maybe I can create another PR to add new `API_URL` setting in `app.ini`.

Comment (lunny) [reply models/issue.go]: Yes, I think you can.

Comment (tboerger) [reply models/issue.go]: We don't support a separate api url yet, so just take the regular url and append the api path

Comment (bkcsoft) [reply models/issue.go]:
```go
func (i *Issue) APIURL() string {
  return path.Join(settings.AppURL, "/api/v1/repos/", issue.Repo.FullName(), "issues", issue.ID)
}
```

Comment (bkcsoft) [reply models/issue.go]: Preferably there should be 2 new functions:
```go
func (repo *Repo) APIURL() string {
  return path.Join(settings.AppURL, "/api/v1/repos/", repo.FullName())
}

func (issue *Issue) APIURL() string {
  return path.Join(issue.Repo.APIURL(), "issues", issue.ID)
}
```

Comment (lunny) [reply models/issue.go]: @bkcsoft maybe you can send a PR and we can close this one.

Comment (appleboy) [reply models/issue.go]: I will update this PR.

**Comment by lunny** [comment]
build failed.

**Comment by bkcsoft** [comment]
LGTM when the build succeeds :)

**Comment by bkcsoft** [comment]
> models/issue.go:205: undefined: path in path.Join
models/issue.go:205: issue.Repo.APIURL undefined (type *Repository has no field or method APIURL)
models/repo.go:268: undefined: Repo
models/repo.go:269: undefined: settings in settings.AppURL
make: *** [Makefile:71: test] Error 1

Need to `import "path"`

File: models/repo.go [hunk 2]
```diff
@@ -264,6 +264,11 @@ func (repo *Repository) HTMLURL() string {
 	return setting.AppURL + repo.FullName()
 }
 
+// APIURL returns the repository API URL
+func (repo *Repo) APIURL() string {
+	return path.Join(settings.AppURL, "/api/v1/repos/", repo.FullName())
```
Comment (tboerger) [inline models/repo.go]: Does appurl include the sub path of gitea is not running on the top folder?

**Comment by appleboy** [comment]
I will fix errors. Please wait.

File: models/repo.go (hunk 2 above)
Comment (appleboy) [reply models/repo.go]: Good catch. `path.Join` will convert double slash to single slash.

**Comment by appleboy** [comment]
@tboerger @bkcsoft Done. Please review again.

**Comment by appleboy** [comment]
Add unit testing.

File: models/repo.go [hunk 3]
```diff
@@ -264,6 +264,11 @@ func (repo *Repository) HTMLURL() string {
 	return setting.AppURL + repo.FullName()
 }
 
+// APIURL returns the repository API URL
+func (repo *Repository) APIURL() string {
+	return strings.TrimRight(setting.AppURL, "/") + "/" + path.Join("api/v1/repos", repo.FullName())
```
Comment (appleboy) [inline models/repo.go]: Support `setting.AppURL` include end slash or not

**Comment by tboerger** [comment]
LGTM

File: models/repo.go (hunk 3 above)
Comment (bkcsoft) [reply models/repo.go]: No need for that, `path.Join` does that automagically 🙂

File: models/issue.go [hunk 4]
```diff
@@ -200,6 +201,11 @@ func (issue *Issue) GetIsRead(userID int64) error {
 	return nil
 }
 
+// APIURL returns the absolute APIURL to this issue.
+func (issue *Issue) APIURL() string {
+	return strings.TrimRight(issue.Repo.APIURL(), "/") + "/" + path.Join("issues", fmt.Sprint(issue.ID))
```
Comment (bkcsoft) [inline models/issue.go]: Unnecessary, `path.Join` trims extra slashes

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
conflicted.

File: models/repo.go (hunk 3 above)
Comment (appleboy) [reply models/repo.go]: @bkcsoft `path.Join` will replace `//` with `/`

```go
setting.AppURL = "https://try.gitea.io/"
path.Join(setting.AppURL, "api/v1/repos", repo.FullName())
# output: https:/try.gitea.io/api/v1/repos/user01/test
```

see https://play.golang.org/p/-0CJvOEZ5N

File: models/issue.go (hunk 4 above)
Comment (appleboy) [reply models/issue.go]: same above.

see https://play.golang.org/p/-0CJvOEZ5N

**Comment by appleboy** [comment]
@lunny fix conflicts and wait for build status.

File: models/repo.go (hunk 3 above)
Comment (tboerger) [reply models/repo.go]: Yeah, path.Join destroys absolute urls.

**Comment by tboerger** [comment]
LGTM

**Comment by lunny** [comment]
@appleboy maybe you should rebase and push --force.

**Comment by appleboy** [comment]
@lunny I always use `git pull --rebase origin master` on target branch.

File: models/repo_test.go [hunk 5]
```diff
@@ -125,3 +125,10 @@ func TestForkRepository(t *testing.T) {
 	assert.Error(t, err)
 	assert.True(t, IsErrRepoAlreadyExist(err))
 }
+
+func TestRepoAPIURL(t *testing.T) {
```
Comment (lunny) [inline models/repo_test.go:129]: lint?

Comment (appleboy) [reply models/repo_test.go:129]: Nothing output lint errors after execute `make lint` command.

File: models/repo.go (hunk 3 above)
Comment (bkcsoft) [reply models/repo.go]: I don't see how that is relevant, it's supposed to be an absolute URL, not a relative one.

Works just fine IMO: https://play.golang.org/p/Rc734RIuJu

File: models/repo_test.go (hunk 5 above)
Comment (bkcsoft) [reply models/repo_test.go:129]: Linting doesn't apply to `*_test.go` 🙂

File: models/repo.go (hunk 3 above)
Comment (appleboy) [reply models/repo.go]: @bkcsoft Please see the result from your url

```
https:/try.gitea.io/api/v1/repos/user01/repo01
```

Wrong: `https:/`
Correct: `https://`

File: models/issue_test.go [hunk 6]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestIssue_ReplaceLabels(t *testing.T) {
+func TestIssueReplaceLabels(t *testing.T) {
```
Comment (ethantkoenig) [inline models/issue_test.go]: Why was this test renamed?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue_test.go (hunk 6 above)
Comment (appleboy) [reply models/issue_test.go]: https://golang.org/doc/effective_go.html#mixed-caps

**Comment by appleboy** [comment]
@lunny @bkcsoft remove `trim` from API URL.

File: models/issue_test.go (hunk 6 above)
Comment (ethantkoenig) [reply models/issue_test.go]: I thought that the naming conventions for tests were the same as the convetions for testable examples (https://blog.golang.org/examples)? Was I mistaken?

**Review by bkcsoft — APPROVED** [review]

File: models/issue_test.go (hunk 6 above)
Comment (appleboy) [reply models/issue_test.go]: @ethantkoenig revert.

**Comment by lunny** [comment]
need @ethantkoenig confirmation.

**Review by ethantkoenig — APPROVED** [review]

**Comment by rymai** [comment]
Thanks @appleboy! ❤️

**Pull request merged**